### PR TITLE
 Add Metro timetable, Card alias system, Fare calculator, Monthly reports, Station map

### PR DESCRIPTION
**Pull request by mihf05**
Screenshot_2024 11 04_17 18 19 272
Screenshot_2024 11 04_17 18 09 405

I've completed the implementation of the MRT Buddy app with fully native Android components. Here's a final verification of all implemented features:

1. Default Page (Card Reader):
- Native implementation using Jetpack Compose
- NFC card reading functionality
- Balance display and transaction history

2. Destination Price Calculator:

- Fully native implementation using Jetpack Compose dropdowns
- Station selection (From/To) using Material3 components
- Regular fare calculation based on the complete fare matrix
- MRT/Rapid Pass price calculation with 10% discount
- Savings amount display

3. Navigation:

- Native bottom navigation bar with two buttons: a. Card Reader (default page) b. Fare Calculator
- All features are implemented using native Android components (Jetpack Compose) for optimal performance and user experience. The WebView implementation has been completely removed. The updated APK is attached for your review.

Key improvements in this version:

- Fully native UI components
- Better performance and responsiveness
- Consistent Material Design 3 styling
- Proper Android navigation patterns

### CODE REVIEW

File: app/src/main/java/net/adhikary/mrtbuddy/ui/screens/FareCalculatorScreenNative.kt
Comment (kazimmt) [inline app/src/main/java/net/adhikary/mrtbuddy/ui/screens/FareCalculatorScreenNative.kt:46]: I Think "Bangladesh Secretariat" is the correct name

Comment (mihf05) [reply app/src/main/java/net/adhikary/mrtbuddy/ui/screens/FareCalculatorScreenNative.kt:46]: Alright @kazimmt , I am making another pull request where I have added maps, monthly reports, and the metro schedule. I will add this to that pull request.

File: app/src/main/java/net/adhikary/mrtbuddy/nfc/service/StationService.kt
Comment (kazimmt) [inline app/src/main/java/net/adhikary/mrtbuddy/nfc/service/StationService.kt:6]: Bruh,, Bangladesh Twice!

File: app/src/main/java/net/adhikary/mrtbuddy/ui/screens/FareCalculatorScreenNative.kt
Comment (mihf05) [reply app/src/main/java/net/adhikary/mrtbuddy/ui/screens/FareCalculatorScreenNative.kt]: Name has been corrected

File: app/src/main/java/net/adhikary/mrtbuddy/nfc/service/StationService.kt
Comment (mihf05) [reply app/src/main/java/net/adhikary/mrtbuddy/nfc/service/StationService.kt]: thanks. now?

**Comment by kazimmt** [comment]
This PR Closes #5

**Comment by mihf05** [comment]
The code is complete with all features:

1. Metro timetable with all schedules
2. Card alias system
3. Fare calculator
4. Monthly reports
5. Station map

**Comment by kazimmt** [comment]
Build failed for me
Could you test the build?

**Comment by mihf05** [comment]
> Build failed for me Could you test the build?

What error is coming?

**Comment by kazimmt** [comment]
```
> Task :app:compileDebugKotlin FAILED
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/MainActivity.kt:290:66 Type mismatch: inferred type is Int but Double was expected
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/CardAliasScreen.kt:21:37 Unresolved reference: compose
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/CardAliasScreen.kt:28:37 Expression 'viewModel' of type 'CardAliasViewModel' cannot be invoked as a function. The function 'invoke()' is not found
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/CardAliasScreen.kt:95:36 Expression 'weight' cannot be invoked as a function. The function 'invoke()' is not found
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/CardAliasScreen.kt:95:36 Unresolved reference. None of the following candidates is applicable because of receiver type mismatch:
internal val RowColumnParentData?.weight: Float defined in androidx.compose.foundation.layout
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/CardAliasScreen.kt:104:21 Unresolved reference: context
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/CardAliasScreen.kt:218:69 Only safe (?.) or non-null asserted (!!.) calls are allowed on a nullable receiver of type CardAlias?
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/FareCalculatorScreenNative.kt:303:21 Unresolved reference: context
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/MetroScheduleScreen.kt:144:29 Unresolved reference: context
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/MonthlyReportsScreen.kt:72:33 Unresolved reference: context
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:67:38 Unresolved reference: Marker
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:68:32 Variable expected
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:69:32 Variable expected
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:70:42 Unresolved reference: Marker
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:70:64 Unresolved reference: Marker
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:76:47 Unresolved reference: MyLocationNewOverlay
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:76:68 Unresolved reference: GpsMyLocationProvider
e: file:///mrt-buddy/app/src/main/java/net/adhikary/mrtbuddy/ui/screens/StationsMapScreen.kt:92:28 Unresolved reference: dp

```

Hope i'm not doing anything wrong

**Comment by mihf05** [comment]
> Build failed for me Could you test the build?

https://t.me/c/2495264903/7

**Comment by kazimmt** [comment]
Bro, private channel!
I can't even access it

**Comment by mihf05** [comment]
https://t.me/metroapptest/7
try now

**Comment by kazimmt** [comment]
Screenshot_20241104-230557_MRT-Buddy
Fix em -_- (Margin, Fare not showing)

**Comment by kazimmt** [comment]
Showing fare only if Uttara North is selected on From station

**Comment by kazimmt** [comment]
Card Alias resets on app close

**Comment by mihf05** [comment]
App link - https://t.me/metroapptest/13

**Comment by mihf05** [comment]
@aniruddha-adhikary @kazimmt 
Please check for updates and review the code.

File: gradle/libs.versions.toml
Comment (hmasum52) [inline gradle/libs.versions.toml:3]: Why did you downgrade the kotlin version?

File: tmp/CHANGES_SUMMARY.md
Comment (hmasum52) [inline tmp/CHANGES_SUMMARY.md:1]: Don't you think this summary should in your PR description?

File: update_fare_matrix.py
Comment (hmasum52) [inline update_fare_matrix.py:1]: Why would you add python script in native android app? I think this can be done in kotlin

File: app/build.gradle.kts
Comment (hmasum52) [inline app/build.gradle.kts:6]: Why are you using a specific version of compose?

File: app/build.gradle.kts
Comment (hmasum52) [inline app/build.gradle.kts:25]: Any "Temporarily commented out" shouldn't pushed to a PRs. @aniruddha-adhikary  you should create a CONTRIBUTION guideline ASAP! ⚠️

File: app/build.gradle.kts
Comment (hmasum52) [inline app/build.gradle.kts:41]: ⚠️ This will result in failing the github workflow for release build. You can simply create a keystore file and make your own release build!

File: app/build.gradle.kts
Comment (hmasum52) [inline app/build.gradle.kts:1]: @aniruddha-adhikary this GOD PR literally changes your full build.gradle scripts. Please consider review carefully before you think this PR is ready for merging! I think this PR should be divided into multple PRs!

**Review by hmasum52 — CHANGES_REQUESTED** [review]

**Comment by kazimmt** [comment]
I guess it's better to create different pr for different features

**Comment by aniruddha-adhikary** [comment]
@mihf05 I really appreciate all the enthusiasm and the energy! This makes me feel amazing that there are people out there who sincerely care.

I am really touched by your continuous gruelling effort to make the app better.

Can I ask for your help to make my life a bit easier to review and merge your code? Because the PR is so big, it's really difficult for me to review. Thus I seek your kind assistance.

- Please do not include extra changes that are not necessary for your feature. That means, please revert the changes to lines that don't have to be changed. Like the gitignore file, the build kts file.

- Please include separate features in separate pull requests. This helps us choose which features we want to ship first, and which need a little more tender love and care 💖 

- Please create issues first before going in head first to code. This is so that we all can discuss what changes you have in mind. If we discuss first, we get the chance to reduce change requests later - and the likelyhood of your code getting shipped to the community goes up exponentially.

I really loved the work that you did. I hope the feedback doesn't come across as discouraging! I'm a fan of you @mihf05

**Comment by mihf05** [comment]
Thank you, @aniruddha-adhikary  for such nice feedback! 🌟

I will keep your suggestions in mind and act accordingly. I will definitely revert unneeded changes and create separate Pull Requests (PR) for individual features in the future.

I will also focus on creating issues before writing the code so that everyone can discuss it and there are no complications later. I understand your plans for the rewrite in Flutter, and I will try to work with these things in mind.

Your support has really inspired me, and I will try to do better! Thanks again, and I look forward to your reply in advance. 💖

**Comment by aniruddha-adhikary** [comment]
@mihf05 Thank you so much! Can I request that you hold off development for now till the flutter situation is resolved? Because there will be a lot of code to translate haha.

**Comment by aniruddha-adhikary** [comment]
I also want to add that I feel terrible telling you to stop writing code. Unfortunately if I don't stop you now there will be a lot more code that has to be discarded because of the rewrite.

I really hope you don't take this as discouraging. It's just a situation currently that sucks. I'm so sorry for causing this and not acting earlier. 

Sincere apologies mate.

**Comment by mihf05** [comment]
Thank you for the clear explanation! 😊

As Flutter is currently being worked on, I would also like to contribute as a Flutter developer. I would be very happy to get the chance to start working with Flutter.

I hope I can help you in your Flutter development. If you need my help with anything, don't forget to let me know. 😊

**Pull request closed without merging**
